### ISSUE TITLE
Dynamic arrays can't be assigned in custom struct arrays

### ISSUE THREAD

**Issue by CisBetter** [adventuregamestudio/ags#338]
### Dynamic arrays can't be assigned in custom struct arrays

When having a custom structure defined as following:

``` AGS-Script
struct Test
{
  InventoryItem * inv_item[];
}
```

you cannot create a `new InventoryItem[i]` if the struct is part of an array itself:

``` AGS-Script
Test my_vector[2];
my_vector.inv_item = new InventoryItem[5];
```
#### The actual error message is:

> Failed to save room room1.crm; details below
> `room1.asc(22): Error (line 22): Type mismatch: cannot convert 'InventoryItem*[]' to 'InventoryItem*'`

**Comment by ivan-mogilko** [comment]
I believe there is a typo in your example, and you meant something like:
<pre>my_vector[0].inv_item = new InventoryItem[5];</pre>
?

Anyway, this looks like a bug related to having those arrays in the struct. Following works:
<pre>
InventoryItem *inv_item[];
inv_item = new InventoryItem[5];
</pre>

Also, if you do 
<pre>my_vector[0].inv_item = new InventoryItem*[5];</pre>
which is not a correct syntax in AGS script, a misleading error message is displayed:
> Built-in type 'InventoryItem' cannot be instantiated directly
(which does not really make sense here)

**Comment by ivan-mogilko** [comment]
Ok, I found more about this. Apparently this bug only occurs when you have array of structs, which have dynamic arrays inside of them.

For instance, if you do:
<pre>
Test single_object;
single_object.inv_item = new InventoryItem[5];
</pre>

That would work. I guess there is a parsing mistake when you do 'arr[N].dynamic_arr' construct.

**Comment by CisBetter** [comment]
Indeed, even though I made a typo I exactly meant this bug :sweat_smile: 

Thanks for your quick reply!
-CisBetter

**Comment by gurok** [comment]
I have been struggling with patching this bug. It seems like it's not just a simple patch. The path resolution part of the compiler isn't built to handle:

StaticArray[index].DynamicArray[index].Value

After modifying the compiler to allow such a construct, I was getting errors like an incorrect area of memory being initialised (null pointer exception when referring to something previously constructed with new).

I am reticent to admit, but I don't fully understand how that part of pointer resolution works yet. I put some logging in the engine to try to understand it, but it's obvious that I'll need more time. More to the point, I think the path resolution might need to be partly rebuilt, and that doesn't make it a simple "patch". The pressure of it being a bug in 3.4.0 that may be holding up the release is also getting to me. I have two alternatives I'd like to suggest:

1. We revert the patch that allowed dynamic arrays inside structs for the release
2. We give a better compiler error that indicates this can't be done yet (perhaps at the point of assignment, or better at the point of declaring such an array of structs -- I don't know if this is as easy)

**Comment by ivan-mogilko** [comment]
Unfortunately, it is not safe to go first way, because this was introduced pretty long ago and some people could already use such thing in their projects. As I pointed out earlier, having dynamic array in single struct object works.
I guess going second way is better in this case - display clear error when user assigns an array (I figure that would be easier to do?) or at the point of declaring array of structs containing managed arrays (if you are confident this will not be too hard to do).
We will also add this to the list of "Known Issues" for now.

**Comment by ivan-mogilko** [comment]
By the way, if the issues you were having were inside the script interpreter, I could look into that too, if you just send me a patch that fixes compiler side.

**Comment by gurok** [comment]
I have resolved this with #343 , but I think this bug should be left open as a feature request. We can look at fixing the path resolution part of the compiler at a later date.

**Comment by ivan-mogilko** [comment]
I have thoroughly investigated that part before, and me and @jjsat also wrote little notes here:
https://github.com/adventuregamestudio/ags/wiki/Script-interpreter
https://github.com/adventuregamestudio/ags/wiki/Script-objects-and-Dynamic-objects

however I realize script interpreter needs better documentation. I will certainly find time to write one.

**Comment by ivan-mogilko** [comment]
> Should we leave this bug open as a feature request and take it off the 3.4.0 milestone

Yes this will be right.

**Comment by ivan-mogilko** [comment]
Regarding scripts, I think you could also use ags disassembler from this repository: https://github.com/rofl0r/agsutils

It takes compiled script file and creates a text file with AGS pseudo-assembler source.

**Comment by ivan-mogilko** [comment]
TL;DR: ags script does not have proper support for complex structs, such as having *arrays of structs that have dynamic arrays inside*.

Using such constructs were temporarily disabled by Gurok (#343).

**Comment by ivan-mogilko** [comment]
Works in the new script compiler in ags4 branch.
